apps/manager_app/user_training.c

```diff
@@ -0,0 +1,285 @@
+
+  evt_config_t evt_config;
+  evt_status_t events;
+  joystick_step_t step[JOYSTICK_TRAIN_STEPS];
+
+  // TODO: add fields for training data if needed
+} training_context_t;
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+
+/**
+ * @brief Initializes a new training context.
+ * @details The function populates the necessary information into the struct
+ * members necessary for the training flow.
+ *
+ * @param query A reference to the query from host for flow initiation
+ *
+ * @return An initialized context with events, config, etc for the user training
+ */
+static training_context_t init_context(manager_query_t *query);
+
+/**
+ * @brief Checks validity of the entry-point training context data.
+ * @details The function will ensure the following requirements:
+ * <ol>
+ * <li> stale p0 events do not exist at entry. </li>
+ * <li> ensure data validity- received query if & is correctly decoded <li>
+ * </ol>
+ * In case of failure, the function will convey appropriate error to host. The
+ * caller function need not handle the usb event as this function does that for
+ * all the early exit cases.
+ *
+ * @return Whether the check is pass or fail.
+ */
+static bool validate_training_request(training_context_t *ctx);
+
+/**
+ * @brief Fill the error in the structure generated via protobuf.
+ */
+static void fill_training_error(manager_train_user_response_t *training,
+                                uint32_t error_code);
+
+/**
+ * @brief Sends the encoded message via usb api to the host.
+ * @details The functions takes the value of fields from the provided reference
+ * to structure instance and encodes into a buffer before using the usb_send_msg
+ * to send data in appropriate format to the host app.
+ *
+ * @param training Reference to a filled structure instance of
+ * manager_train_user_response_t
+ */
+static void send_msg_to_host(manager_train_user_response_t *training);
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+static training_context_t init_context(manager_query_t *query) {
+  training_context_t context = {
+      .training_resp = MANAGER_TRAIN_USER_RESPONSE_INIT_ZERO,
+      .training_req = query,
+      .evt_config = {.timeout = 3000, .evt_selection = EVENT_CONFIG_USB},
+      .events = {{0}, {0}, {0}, {0}},
+      .step = {{.instruction = "Toggle up your joystick",
```

Comment:
Let's keep constant texts in a single place.. and use pointers to use them

Reply:
moved to `constants_text.c`